File: src/strategies/funding.js
// nexus/src/strategies/funding.js — Funding Rate Harvest Strategy
//
// When a perpetual's funding rate is sufficiently high (positive or negative),
// holding a delta-neutral position (long spot + short perp, or vice-versa)
// generates risk-free yield equal to the collected funding payment minus
// round-trip fees.  This strategy surfaces such opportunities for the
// orchestrator to act on.

// Minimum absolute funding rate to consider (per 8-hour settlement period).
// 0.0001 = 0.01% per period ≈ 10.95% APY — below this the edge is marginal.
const MIN_FUNDING_RATE = 0.0001;

// Cap: funding rates above 1% per period usually revert before settlement.
const MAX_FUNDING_RATE = 0.01;

// Maximum allowed spot-vs-perp price divergence before the spread guard fires.
// Above this the mark price is unreliable and execution slippage would eat profit.
const MAX_DIVERGENCE_PCT = 2.0;

/**
 * Evaluates whether a funding-rate harvest trade is worthwhile.
 *
 * @param {string}      symbol        — e.g. 'BTCUSDT'
 * @param {Array}       spotSources   — [{ price, exchange, fee }]
 * @param {object|null} perpData      — { price, exchange, fee, fundingRate }
 * @param {number}      maxSpreadPct  — additional volatility guard from config
 * @returns {object|null}  OpportunityObject or null
 */
export function scanFundingRate(symbol, spotSources, perpData, maxSpreadPct) {
  if (!perpData || spotSources.length < 1) return null;

  const fundingRate = perpData.fundingRate ?? 0;
  const absFunding  = Math.abs(fundingRate);

  if (absFunding < MIN_FUNDING_RATE) return null;
  if (absFunding > MAX_FUNDING_RATE) return null;

  // Use cheapest spot source to minimise entry cost
  const bestSpot = spotSources.reduce((a, b) => (a.price < b.price ? a : b));

  // Reject if spot and perp price diverge suspiciously (stale data / illiquid)
  const divergencePct = Math.abs(perpData.price - bestSpot.price) / bestSpot.price * 100;
  if (divergencePct > Math.min(maxSpreadPct, MAX_DIVERGENCE_PCT)) return null;

  // P&L per period = |fundingRate| - round-trip taker fees (enter + exit, both legs)
  const roundTripFeePct = (bestSpot.fee + perpData.fee) * 2 * 100;
  const fundingPct      = absFunding * 100;
  const netPct          = fundingPct - roundTripFeePct;

  if (netPct <= 0) return null;

  if (fundingPct === 0) return null;
  const safetyFactor = netPct / fundingPct;

  // Positive funding → shorts receive payment → go long spot + short perp
  // Negative funding → longs receive payment  → go short spot + long perp
  const receiveFunding = fundingRate >= 0; // true = receive as short perp
  const buyExchange  = receiveFunding ? bestSpot.exchange : perpData.exchange;
  const sellExchange = receiveFunding ? perpData.exchange : bestSpot.exchange;
  const buyPrice     = receiveFunding ? bestSpot.price    : perpData.price;
  const sellPrice    = receiveFunding ? perpData.price    : bestSpot.price;
  const perpSide     = receiveFunding ? 'SHORT'           : 'LONG';

  return {
    strategy:       'funding',
    symbol,
    buyExchange,
    sellExchange,
    buyPrice,
    sellPrice,
    grossPct:       fundingPct,
    netPct,
    safetyFactor,
    direction:      `SPOT→${perpData.exchange.toUpperCase()} ${perpSide} (funding harvest)`,
    isPerp:         true,
    fundingRate,
    perpSide,
    fundingHarvest: true
  };
}

File: hft/internal/strategies/funding/funding.go
// Package funding implements the funding-rate harvest strategy.
// When a perpetual's 8-hour funding rate is large enough to cover round-trip
// fees, holding a delta-neutral position (long spot + short perp, or vice-versa)
// generates risk-free yield.
package funding

import (
	"fmt"
	"math"

	"github.com/zedanazad43/UltimateArbitrageHFT/hft/internal/feeds"
	"github.com/zedanazad43/UltimateArbitrageHFT/hft/internal/strategies/cex"
)

const (
	// minFundingRate is the minimum |fundingRate| to consider.
	// 0.0001 = 0.01% per 8-hour period ≈ 10.95% APY.
	minFundingRate = 0.0001
	// maxFundingRate caps rates that are likely to revert before settlement.
	maxFundingRate = 0.01
	// maxDivergencePct caps the spot-vs-perp price divergence.
	maxDivergencePct = 2.0
)

// Opportunity extends cex.Opportunity with funding-specific fields.
type Opportunity struct {
	*cex.Opportunity
	FundingRate    float64
	PerpSide       string // "LONG" or "SHORT"
	FundingHarvest bool
}

// Scan evaluates whether a funding-rate harvest trade is worthwhile.
// It returns nil when there is no actionable opportunity.
func Scan(symbol string, spotSources []feeds.PriceSource, perpData *feeds.PerpData, maxSpreadPct float64) *Opportunity {
	if perpData == nil || len(spotSources) == 0 {
		return nil
	}

	fundingRate := perpData.FundingRate
	absFunding := math.Abs(fundingRate)

	if absFunding < minFundingRate || absFunding > maxFundingRate {
		return nil
	}

	// Use cheapest spot source.
	bestSpot := spotSources[0]
	for _, s := range spotSources[1:] {
		if s.Price < bestSpot.Price {
			bestSpot = s
		}
	}

	// Reject if spot and perp price diverge suspiciously.
	divergencePct := math.Abs(perpData.Price-bestSpot.Price) / bestSpot.Price * 100
	limit := math.Min(maxSpreadPct, maxDivergencePct)
	if divergencePct > limit {
		return nil
	}

	// P&L per period = |fundingRate| - round-trip taker fees (enter + exit, both legs).
	roundTripFeePct := (bestSpot.Fee + perpData.Fee) * 2 * 100
	fundingPct := absFunding * 100
	netPct := fundingPct - roundTripFeePct

	if netPct <= 0 || fundingPct == 0 {
		return nil
	}
	safetyFactor := netPct / fundingPct

	// Positive funding → shorts receive payment → go long spot + short perp.
	// Negative funding → longs receive payment  → go short spot + long perp.
	receiveFunding := fundingRate >= 0
	perpSide := "LONG"
	if receiveFunding {
		perpSide = "SHORT"
	}
	buyExchange, sellExchange := perpData.Exchange, bestSpot.Exchange
	buyPrice, sellPrice := perpData.Price, bestSpot.Price
	if receiveFunding {
		buyExchange, sellExchange = bestSpot.Exchange, perpData.Exchange
		buyPrice, sellPrice = bestSpot.Price, perpData.Price
	}

	return &Opportunity{
		Opportunity: &cex.Opportunity{
			Strategy:     "funding",
			Symbol:       symbol,
			BuyExchange:  buyExchange,
			SellExchange: sellExchange,
			BuyPrice:     buyPrice,
			SellPrice:    sellPrice,
			GrossPct:     fundingPct,
			NetPct:       netPct,
			SafetyFactor: safetyFactor,
			Direction:    fmt.Sprintf("SPOT→%s %s (funding harvest)", perpData.Exchange, perpSide),
			IsPerp:       true,
		},
		FundingRate:    fundingRate,
		PerpSide:       perpSide,
		FundingHarvest: true,
	}
}
